staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go

```diff
@@ -146,14 +146,29 @@ type Store struct {
 	// specific cases where storage of the value is not appropriate, since
 	// they cannot be watched.
 	Decorator ObjectFunc
+
 	// CreateStrategy implements resource-specific behavior during creation.
 	CreateStrategy rest.RESTCreateStrategy
+	// BeginCreate is an optional hook that returns a "transaction-like"
+	// commit/cleanup function which will be called at the end of operation,
+	// whether it succeeds or not.  If the overall operation panics, the
+	// cleanup function will not be called.  Almost nobody should use this
+	// hook.
+	//FIXME: do we need to pass obj to cleanup?
+	BeginCreate func(obj runtime.Object, options *metav1.CreateOptions) (func(runtime.Object, error), error)
 	// AfterCreate implements a further operation to run after a resource is
 	// created and before it is decorated, optional.
 	AfterCreate ObjectFunc
 
 	// UpdateStrategy implements resource-specific behavior during updates.
 	UpdateStrategy rest.RESTUpdateStrategy
+	// BeginUpdate is an optional hook that returns a "transaction-like"
+	// commit/cleanup function which will be called at the end of operation,
+	// whether it succeeds or not.  If the overall operation panics, the
+	// cleanup function will not be called.  Almost nobody should use this
+	// hook.
+	//FIXME: do we need to pass objs to cleanup?
+	BeginUpdate func(obj, old runtime.Object, options *metav1.UpdateOptions) (func(runtime.Object, runtime.Object, error), error)
```

Comment:
In a patch flow, when does this get called?  I would expect it to be after the patch is applied every time.

Reply:
Yes.  Patches are applied ca L528, I think:

```
obj, err := objInfo.UpdatedObject(ctx, existing)
```

The `obj` argument to BeginUpdate is `obj`, and the `old` argument is `existing`

In case of a conflict, this loop (as I understand it) will re-fetch `existing` and re-apply the patch (assuming preconditions pass).